src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertForEachToLinqQueryTests.cs

```diff
@@ -0,0 +1,2842 @@
+        }|]
+    }
+}
+";
+            string output = @"
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+static class Extensions
+{
+    public static IEnumerable<C> Select(this int[] x, Func<int, C> predicate) => throw null;
+}
+
+class C
+{
+    public static implicit operator int(C x)
+    {
+        throw null;
+    }
+
+    public static implicit operator C(int x)
+    {
+        throw null;
+    }
+
+    IEnumerable<C> Test()
+    {
+        return from x in new[] { 1, 2, 3 }
+               select x;
+    }
+}
+";
+            await TestAsync(source, output, parseOptions: null);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task IQueryableConvertedToIEnumerableInReturn()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+
+class C
+{
+    IEnumerable<int> M(IEnumerable<int> nums)
+    {
+        [|foreach (int n1 in nums.AsQueryable())
+        {
+            yield return n1;
+        }|]
+
+        yield break;
+    }
+}";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+
+class C
+{
+    IEnumerable<int> M(IEnumerable<int> nums)
+    {
+        return from int n1 in nums.AsQueryable()
+               select n1;
+    }
+}";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ReturnIQueryableConvertedToIEnumerableInAssignment()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+
+class C
+{
+    IEnumerable<int> M(IEnumerable<int> nums)
+    {
+        [|foreach (int n1 in nums.AsQueryable())
+        {
+            yield return n1;
+        }|]
+    }
+}";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+
+class C
+{
+    IEnumerable<int> M(IEnumerable<int> nums)
+    {
+        return from int n1 in nums.AsQueryable()
+               select n1;
+    }
+}";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        #endregion
+
+        #region In ToList
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListLastDeclarationMerge()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        List<int> list0 = new List<int>(), list = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                list.Add(n1);
+            }
+        }|]
+
+        return list;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        List<int> list0 = new List<int>();
```

Comment:
is this merging only works for known type like List I guess? or only List?

Reply:
if we already special case List, I guess we can remove from int n2 in nums if we know it has no side effect (nums) and n2 is never used?

Reply:
I think it would be hard to determine if there are no side effects. It seems to be a theme for a separate analyzer/refactoring.

---
In reply to: [195855642](https://github.com/dotnet/roslyn/pull/27130#discussion_r195855642) [](ancestors = 195855642)

Reply:
correct, only List is supported

---
In reply to: [195855452](https://github.com/dotnet/roslyn/pull/27130#discussion_r195855452) [](ancestors = 195855452)